### DISCUSSION TITLE
Use Solargraph dockerized with ruby files

### DISCUSSION THREAD

**Discussion by mattszein** [LazyVim/LazyVim#1990] [Q&A]
Hi! I need help, i am new using lazyvim and nvim.
I always use a dockerized environment to develop in ruby, i mean i dont have ruby installed in my host system, i use containers with docker.
I want to use solargraph and it is running in a container, but i dont know how to connect it with nvim to use it. 
I know i could use it installing Coc and coc-solargraph with the externalserver config but i am trying to configure with nvim-lspconfig.

I have this file called solargraph.lua:
```
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      solargraph = {},
    },
    setup = {
      solargraph = function(_, opts)
        require("lspconfig").solargraph.setup({
          cmd = { "docker", "compose", "up", "solargraph" },
          root_dir = nvim_lsp.util.root_pattern("Gemfile", ".git", "."),
          settings = {
            solargraph = {
              autoformat = false,
              formatting = false,
              completion = true,
              diagnostic = true,
              folding = true,
              references = true,
              rename = true,
              symbols = true,
            },
          },
        })
        return true
      end,
      -- Specify * to use this function as a fallback for any server
      -- ["*"] = function(server, opts) end,
    },
  },
}
```

I have an error:
```
`[INFO  Fri 10 Nov 2023 08:05:18 PM EET] ...e/nvim/lazy/mason.nvim/lua/mason-core/installer/init.lua:184: Executing installer for Package(name=solargraph) {}
[ERROR Fri 10 Nov 2023 08:05:20 PM EET] ...e/nvim/lazy/mason.nvim/lua/mason-core/installer/init.lua:249: Installation failed for Package(name=solargraph) error=spawn: gem failed with exit code 1 and signal 0. `
```

**Comment by dpetka2001** [comment]
It seems that you need `gem` installed to be able to install solargraph via mason. I don't have experience with this, I'm just guessing from the error you posted. On Ubuntu this could be done with `sudo apt install rubygems ruby-dev` from what I googled around.

**Reply by mattszein** [reply]
Yes. but the idea is avoid install ruby and a specific Gemfile in my computer. I use docker in development to avoid install dependencies in my computer.
Is there a way to use a lsp server without installing the gem in my computer using nvim-lspconfig?
How can i avoid the mason installation related to solargraph?

**Reply by dpetka2001** [reply]
Mason uses gem to install the LSP server. You would have to install it yourself and make it available in your PATH for nvim-lspconfig to pick it up. To avoid mason installation you can do 
```lua
opts = {
    servers = {
      solargraph = {
	  	mason = false,
	  },
    },
}
```
Then mason won't install it and lspconfig should pick up your the LSP server that you installed yourself.

**Reply by mattszein** [reply]
Great!
Thank you. It's working.
My config is:
```
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      solargraph = {
        mason = false,
      },
    },
    setup = {
      solargraph = function(_, opts)
        require("lspconfig").solargraph.setup({
          cmd = { "docker", "compose", "run", "solargraph", "solargraph", "stdio" },
          settings = {
            solargraph = {
              autoformat = false,
              formatting = false,
              completion = true,
              diagnostic = true,
              folding = true,
              references = true,
              rename = true,
              symbols = true,
            },
          },
        })
        return true
      end,
    },
  },
}
```
